robot/connection.py

```diff
@@ -0,0 +1,42 @@
+import json
+import socket
+from typing import Any, Dict
+
+Message = Dict[str, Any]
+
+
+class Connection:
+    """
+    A connection to a device.
+
+    This wraps a ``socket.socket`` providing encoding and decoding so that
+    consumers of this class can send and receive JSON-compatible typed data
+    rather than needing to worry about lower-level details.
+    """
+
+    def __init__(self, socket: socket.socket) -> None:
+        """Wrap the given socket."""
+        self.socket = socket
+        self.data = b''
+
+    def close(self) -> None:
+        """Close the connection."""
+        self.socket.close()
+
+    def send(self, message: Message) -> None:
+        """Send the given JSON-compatible message over the connection."""
+        line = json.dumps(message).encode('utf-8') + b'\n'
+        self.socket.sendall(line)
+
+    def receive(self) -> Message:
+        """Receive a single message from the connection."""
+        while b'\n' not in self.data:
+            message = self.socket.recv(4096)
+            if message == b'':
+                raise BrokenPipeError()
+
+            self.data += message
```

Comment:
Would it be worth using an [`io.BufferedReader`](https://docs.python.org/3/library/io.html#io.BufferedReader) here instead of implementing buffering ourselves? I haven't used this particular class before, but my intuition suggests we can just call `readline` on an instance of `BufferedReader` and it will deal with performing read operations on the socket only when needed.

However, your current approach has the advantage of being more explicit about its error handling. It's not clear without looking at its source how `BufferedReader` would handle an empty response from `socket.recv`.

Reply:
Interesting, I'll give that a look. I also found that we can turn a socket into a file-like object, though again I'm not sure how that handles errors.

As it stands, this class is pretty much copied from [robotd/master.py](https://github.com/sourcebots/robotd/blob/e0be841816d4c778e98498193f6fd7e3de35813d/robotd/master.py#L19-L53). If we do find that there's a good wrapper then we should consider using it there, too.

Reply:
Tangentially related (and definitely out of scope), but a long-term improvement could be to factor out this logic from both robotd and robot-api into a library used by both (e.g. "robot-protocol"). This would facilitate writing other applications that interact with robotd, such as a manual control interface or the fabled Power Board Orchestra.

Reply:
Yup, a common library was another thing I'd wondered about here. I think we'd want to pin down some even higher level things before/while doing that (such as a generic way to pass errors back & forth) too.

I'd hope that something like a manual control interface would sit above the robot-api though -- is there a reason you think it needs to be a sibling?

Reply:
>I'd hope that something like a manual control interface would sit above the robot-api though

My bad - that would make much more sense that using the hypothetical `robot-protocol` package directly.

Come to think of it I'm not sure what applications, if any, *would* use `robot-protocol` rather than `robot-api` in this sense.